### PR TITLE
counsel: add tab switcher

### PR DESCRIPTION
**Pull request by peterhoeg**
Allows you to easily switch between open tabs.

### CODE REVIEW

**Comment by peterhoeg** [comment]
CA signed ages ago due to `use-package` contribution.

File: counsel.el [hunk 1]
```diff
@@ -731,6 +731,25 @@ With a prefix arg, restrict list to variables defined using
       (when doc
         (lv-delete-window)))))
 
+;;** `counsel-tab-bar'
+(defun counsel--tab-name (tab)
+  "Return a name for a tab."
+  (cdr (car (cdr tab))))
```
Comment (basil-conto) [inline counsel.el:737]: This is assuming that `name` is the first element of a tab's alist, but alists are unordered collections. You should instead write:
```el
(cdr (assq 'name tab))
```

File: counsel.el [hunk 2]
```diff
@@ -731,6 +731,25 @@ With a prefix arg, restrict list to variables defined using
       (when doc
         (lv-delete-window)))))
 
+;;** `counsel-tab-bar'
```
Comment (basil-conto) [inline counsel.el:734]: This section is under `;;* Elisp symbols`, where it does not belong. Please move it under `;; Misc. Emacs` or similar.

File: counsel.el [hunk 3]
```diff
@@ -731,6 +731,25 @@ With a prefix arg, restrict list to variables defined using
...
+(defun counsel--tab-name (tab)
+  "Return a name for a tab."
+  (cdr (car (cdr tab))))
+
+;;;###autoload
+(defun counsel-tab-bar ()
+  "Switch between tabs."
+  (interactive)
+  (ivy-read
+   "Select tab: "
```
Comment (basil-conto) [inline counsel.el:744]: Nit: perhaps the prompt should convey what `counsel-tab-bar` is going to do with the tab, e.g. `Switch to tab:`.

File: counsel.el [hunk 4]
```diff
@@ -731,6 +731,25 @@ With a prefix arg, restrict list to variables defined using
...
+   (mapcar #'counsel--tab-name (funcall tab-bar-tabs-function))
+   :Keymap counsel-mode-map
```
Comment (basil-conto) [inline counsel.el:746]: `:Keymap` -> `:keymap`

But I don't think this is right, as `counsel-mode-map` is enabled by `counsel-mode`, not individual completion functions.

File: counsel.el [hunk 5]
```diff
@@ -731,6 +731,25 @@ With a prefix arg, restrict list to variables defined using
...
+  "Return a name for a tab."
+  (cdr (car (cdr tab))))
+
+;;;###autoload
+(defun counsel-tab-bar ()
+  "Switch between tabs."
+  (interactive)
+  (ivy-read
+   "Select tab: "
+   (mapcar #'counsel--tab-name (funcall tab-bar-tabs-function))
```
Comment (basil-conto) [inline counsel.el:745]: `tab-bar-tabs-function` and `tab-bar-switch-to-tab` are new in Emacs 27, so in older Emacsen they give byte-compiler warnings like the following:
```
In counsel-tab-bar:
counsel.el:745:41:Warning: reference to free variable ‘tab-bar-tabs-function’

In end of data:
counsel.el:6965:1:Warning: the following functions are not known to be
    defined: tab-bar-switch-to-tab
```
I think the best way to silence these is by adding the following after `(interactive)`:
```el
  (when (< emacs-major-version 27)
    (user-error "This command requires Emacs 27 or later"))
  (defvar tab-bar-tabs-function)
  (declare-function tab-bar-switch-to-tab "tab-bar" (name))
```

File: counsel.el [hunk 6]
```diff
@@ -731,6 +731,25 @@ With a prefix arg, restrict list to variables defined using
...
+   :Keymap counsel-mode-map
+   :preselect (counsel--tab-name (tab-bar--current-tab))
```
Comment (basil-conto) [inline counsel.el:747]: Please avoid calling internal functions like `tab-bar--current-tab`, as they can come and go and change behaviour without warning. Instead, can't you use the `current-tab` entry returned by `tab-bar-tabs-function`?

Stepping even further back, why preselect the current tab? Doesn't switching to a tab imply that the user wants to switch to a different tab?

File: counsel.el [hunk 7]
```diff
@@ -731,6 +731,25 @@ With a prefix arg, restrict list to variables defined using
...
+   :require-match t
+   :action (lambda (x)
+             (tab-bar-switch-to-tab x))
```
Comment (basil-conto) [inline counsel.el:750]: AKA `:action #'tab-bar-switch-to-tab`

**Review by basil-conto — CHANGES_REQUESTED** [review]
Thanks for working on this! See some comments on technicalities below, otherwise LGTM.

**Comment by peterhoeg** [comment]
@basil-conto my apologies for dropping this after you took the time to review it. I am no longer using counsel, so closing this again.

**Pull request closed without merging**